GeneratorInterface/RivetInterface/plugins/RivetAnalyzer.cc

```diff
@@ -133,7 +133,7 @@ void RivetAnalyzer::analyze(const edm::Event& iEvent,const edm::EventSetup& iSet
   _analysisHandler.analyze(*myGenEvent);
 
   //if we have cloned the GenEvent, we delete it
-  if ( _useExternalWeight ) 
+  if ( _useExternalWeight || _xsection > 0 ) 
```

Comment:
It would be better to add at line 92
```
std::unique_ptr<HepMC::GenEvent> tmpGenEvtPtr;
```
Then replace line 94 with
```
tmpGenEvtPtr = std::make_unique<HepMC::Event>(*(evt->GetEvent()));
```

Then line 121 becomes
```
myGenEvent = tmpGenEvtPtr.get();
```
Finally you then can get rid of lines 136 and 137. 

This change also avoids a memory leak if the code throws at line  104.